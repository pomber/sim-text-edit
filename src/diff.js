import { diffChars, diffWordsWithSpace } from "diff";

export function diff(prev, next) {
  const changes = diffWordsWithSpace(prev, next, { newlineIsToken: true });
  let newChanges = [];
  changes.forEach((change) => {
    if (change.removed) {
      const [firstDeletion, ...linesRemoved] = change.value.split("\n");
      newChanges.push({
        value: firstDeletion,
        removed: change.removed,
        added: change.added,
        count: firstDeletion.length,
      });

      newChanges = newChanges.concat(
        linesRemoved.map((lineRemoved) => ({
          value: "\n" + lineRemoved,
          removed: change.removed,
          added: change.added,
          count: lineRemoved.length + 1,
        }))
      );
    } else if (change.added) {
      const linesAdded = change.value.split("\n");
      const lastAddition = linesAdded.pop();

      newChanges = newChanges.concat(
        linesAdded.map((line) => ({
          value: line + "\n",
          removed: change.removed,
          added: change.added,
          count: linesAdded.length + 1,
        }))
      );

      newChanges.push({
        value: lastAddition,
        removed: change.removed,
        added: change.added,
        count: lastAddition.length,
      });
    } else {
      newChanges.push(change);
    }
  });
  return newChanges;
}
