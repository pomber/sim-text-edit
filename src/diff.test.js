import { diff } from "./diff";

it("works", () => {
  const result = diff("a", "b");
  expect(result).toMatchSnapshot();
});
