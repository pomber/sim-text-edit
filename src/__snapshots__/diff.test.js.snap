// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`works 1`] = `
Array [
  Object {
    "added": undefined,
    "count": 1,
    "removed": true,
    "value": "a",
  },
  Object {
    "added": true,
    "count": 1,
    "removed": undefined,
    "value": "b",
  },
]
`;
